src/main/java/com/minecolonies/coremod/colony/buildings/inventory/ItemStackHandlerWithIndex.java

```diff
@@ -0,0 +1,333 @@
+            {
+                meta.removeFromIndex(store, metaIndexValue);
+            }
+            usedSlots--;
+        }
+
+        // Some operations on forge side do not use extractitem cleanly, but instead use it to simulate and
+        // then set the slot to empty via this call, so gotta support setting empty stacks.
+        if (!ItemStackUtils.isEmpty(storage.getItemStack()))
+        {
+            // Adding stack to indexes
+            index.addToIndex(storage, slot);
+
+            if (meta != null)
+            {
+                meta.addToIndex(storage, metaIndexValue);
+            }
+            usedSlots++;
+        }
+        super.setStackInSlot(slot, storage.getItemStack());
+    }
+
+    @Override
+    @Nonnull
+    public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate)
+    {
+        if (!isValidSlot(slot) || ItemStackUtils.isEmpty(stack))
+        {
+            return stack;
+        }
+
+        int insert = getStackLimit(slot, stack);
+
+        if (getStackInSlot(slot) != ItemStack.EMPTY)
+        {
+            if (!getStackInSlot(slot).equals(stack))
+            {
+                return stack;
+            }
+            insert -= getStackInSlot(slot).getCount();
+        }
+
+        // The actual amount we're inserting.
+        insert = Math.min(insert, stack.getCount());
+
+        if (insert <= 0)
+        {
+            return stack;
+        }
+
+        // Return the remainder right away when simulating
+        if (simulate)
+        {
+            ItemStack returnStack = stack.copy();
+            returnStack.setCount(stack.getCount() - insert);
+            return returnStack;
+        }
+
+        // Modify the inventory contents
+        if (getStackInSlot(slot) != ItemStack.EMPTY)
+        {
+            getStackInSlot(slot).grow(insert);
+            onContentsChanged(slot);
+
+            if (stack.getCount() - insert <= 0)
+            {
+                return ItemStack.EMPTY;
+            }
+
+            // Return the remainder
+            ItemStack remainder = stack.copy();
+            remainder.setCount(stack.getCount() - insert);
+
+            return remainder;
+        }
+        else
+        {
+            setStackInSlot(slot, stack);
+            return ItemStack.EMPTY;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public ItemStack extractItem(final int slot, final int amount, final boolean simulate)
+    {
+        if (!isValidSlot(slot) || ItemStackUtils.isEmpty(getStackInSlot(slot)))
+        {
+            return ItemStack.EMPTY;
+        }
+
+
+        int toExtract = Math.min(amount, getStackInSlot(slot).getCount());
+
+        if (simulate)
+        {
+            ItemStack simulateStack = getStackInSlot(slot).copy();
+            simulateStack.setCount(toExtract);
+            return simulateStack;
+        }
+
+        ItemStack returnStack;
+        // Returning whole stack
+        if (toExtract >= getStackInSlot(slot).getCount())
```

Comment:
This should be an equal.....

Reply:
This whole if branch can be done in about three lines, with a copy of the itemstack, a shrink of the original one and a set size on the copy stack who is then returned........